src/relentless/simulate/dilute.py

```diff
@@ -200,12 +198,31 @@ def process(self, sim, sim_op):
                     raise ValueError(
                         "Geometric integration factor unknown for extent type"
                     )
-                y = (geo_prefactor / 6.0) * rho_a * rho_b * f * gr * r
-                ens.P += numpy.trapz(y, x=r)
-
-        sim[self]["ensemble"] = ens
-        sim[self]["num_thermo_samples"] = None
-        sim[self]["num_rdf_samples"] = None
+                rho = N / ens.V.extent
+                y_a = ens.N[a] / N
+                y_b = ens.N[b] / N
+                b_ab = (geo_prefactor / 2.0) * (numpy.exp(-u / kT) - 1)
+                B += y_a * y_b * numpy.trapz(b_ab, x=r)
+        # calculate pressure if no barostat
+        if ens.P == 0.0:
+            ens.P = kT * (rho + B * rho**2)
+            sim[self]["ensemble"] = ens
+            sim[self]["num_thermo_samples"] = None
+            sim[self]["num_rdf_samples"] = None
+        # adjust extent to maintain pressure if barostat
+        else:
+            coeffs = numpy.array([-ens.P / kT, N, B * N**2])
```

Comment:
Are these the right coefficients? I don't have pencil & paper in front of me, but I think the virial expansion is:
```
P/kT = rho + B rho^2
```
Then solve as:
```
-P/kT + rho + B rho^2 = 0
```
You could also rearrange for the molar volume `v = 1/rho`:
```
B + v + (-P/kT) v^2 = 0
```
Then you would compute `V = N/rho` or `V = N*v`.

Right now, I think you are actually getting `1/V` and not `V`.

Reply:
My understanding is:
`rho = N/V`
&
`-P/kT + rho + B rho^2 = 0`



Which gives: 
`-P/kT + N/V + B N^2 /V^2 = 0` 

Multiplying `V^2` gives:
`(-P/kT)*V^2 + N*V + BN^2 =0`

The length value output gives the right pressure when running the simulation with no barostat.

Reply:
Oh! I see, the problem is that I misread the order of coefficients for roots. I thought they would go in ascending order (a0 + a1 c + a2 x^2 + …), but they are descending.

For numerics, you may still want to solve for molar volume or density, so that the variable you are solving for is not extensive.